ember-server/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,217 @@
+  private[this] val webSocketCi = "websocket".ci
+  private[this] val connectionUpgrade = Connection(NonEmptyList.of(upgradeCi))
+  private[this] val upgradeWebSocket = Upgrade(NonEmptyList.of(webSocketCi))
+
+  // TODO: Express this in terms of Stream to leverage interrupt machinery
+  def upgrade[F[_]](
+      socket: Socket[F],
+      req: Request[F],
+      ctx: WebSocketContext[F],
+      buffer: Array[Byte],
+      receiveBufferSize: Int,
+      idleTimeout: Duration,
+      onWriteFailure: (Option[Request[F]], Response[F], Throwable) => F[Unit])(implicit
+      F: Concurrent[F]): F[Unit] = {
+    val wsResponse = clientHandshake(req) match {
+      case Right(key) =>
+        serverHandshake(key)
+          .map { accept =>
+            val secWebSocketAccept = `Sec-WebSocket-Accept`(accept)
+            val headers =
+              ctx.headers ++ Headers.of(connectionUpgrade, upgradeWebSocket, secWebSocketAccept)
+            Response[F](Status.SwitchingProtocols)
+              .withHeaders(headers)
+          }
+          .handleError(_ =>
+            Response[F](Status.InternalServerError).withEntity(
+              "Encountered an error during WebSocket handshake."))
+      case Left(error) =>
+        // TODO: insert the appropriate headers
+        Response[F](error.status).withEntity(error.message).pure[F]
+    }
+
+    wsResponse
+      .flatMap { res =>
+        ServerHelpers.send(socket)(Some(req), res, idleTimeout, onWriteFailure).void
+      } >> runConnection(socket, ctx, buffer, receiveBufferSize, idleTimeout)
+  }
+
+  private def runConnection[F[_]](
+      socket: Socket[F],
+      ctx: WebSocketContext[F],
+      buffer: Array[Byte],
+      receiveBufferSize: Int,
+      idleTimeout: Duration)(implicit F: Concurrent[F]): F[Unit] = {
+    val read: Read[F] = socket.read(receiveBufferSize, durationToFinite(idleTimeout))
+    val frameTranscoder = new FrameTranscoder(false)
+
+    // TODO: make sure error semantics are correct and that resources are properly cleaned up
+    // TODO: consider write/read failures and effect on outer connection
+    // TODO: there is some shared code here with ServerHelpers
+    val writer = ctx.webSocket.send
+      .flatMap { frame =>
+        // TODO: frameToBuffer can throw
+        Stream
+          .iterable(frameTranscoder.frameToBuffer(frame).map { buffer =>
+            // TODO: improve
+            val bytes = new Array[Byte](buffer.remaining())
+            buffer.get(bytes)
+            Chunk.bytes(bytes)
+          })
+          .flatMap(Stream.chunk(_))
+      }
+      .through(socket.writes(durationToFinite(idleTimeout)))
+
+    val reader = (Stream.chunk(Chunk.bytes(buffer)) ++ readStream(read))
+      .through(decodeFrames(frameTranscoder))
+      .through(ctx.webSocket.receive)
+
+    reader
+      .concurrently(writer)
+      .drain
+      .compile
+      .drain
+      .attempt
+      .flatMap {
+        case Left(err) =>
+          err.printStackTrace()
```

Comment:
We're missing some loggers here and on the println.

Reply:
I think these will just go away at the end. I haven't made a habit of adding debug log statements; is that something we want to do more of?